<!-- sct-test-summary -->
## :x: Test Summary: FAILED

### :white_check_mark: Precommit: PASSED

| Total | Passed | Failed | Skipped |
|-------|--------|--------|---------|
| 26 | 15 | 0 | 11 |


### :x: Tests: FAILED

| Total | Passed | Failed | Errors | Skipped |
|-------|--------|--------|--------|---------|
| 1798 | 1777 | 6 | 0 | 15 |

### Failed Tests

#### `unit.test_cluster_oci.test_private_dns_name_uses_metadata_hostname_when_available`
**Type:** FAILURE
**Message:** `TypeError: _base_node_init() got an unexpected keyword argument 'after_config'`
<details><summary>Traceback</summary>

```
mock_resolver = <MagicMock name='resolve_ip_to_dns' id='136917591462576'>

    @patch("sdcm.cluster_oci.resolve_ip_to_dns", return_value="node.internal")
    @patch("sdcm.cluster.BaseNode.__init__", new=_base_node_init)
    def test_private_dns_name_uses_metadata_hostname_when_available(mock_resolver) -> None:
        """Test that FQDN from OCI metadata is used directly without reverse DNS lookup."""
>       oci_node = OciNode(_mock_cloud_instance(), MOCK_CREDENTIALS, MOCK_PARENT_CLUSTER)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

unit_tests/unit/test_cluster_oci.py:38: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <sdcm.cluster_oci.OciNode object at 0x7c869c671550>
oci_instance = <Mock id='136917591462912'>
credentials = <Mock id='136917979654400'>
parent_cluster = <Mock id='136917979654736'>, node_prefix = 'node'
node_index = 1, base_logdir = None, dc_idx = 0, rack = 0, after_config = None

    def __init__(
        self,
        oci_instance: VmInstance,
        credentials,
        parent_cluster,
        node_prefix="node",
        node_index=1,
        base_logdir=None,
        dc_idx=0,
        rack=0,
        after_config=None,
    ):
        self.node_index = node_index
        self.dc_idx = dc_idx
        self.parent_cluster = parent_cluster
        self._instance = oci_instance
        self._instance_type = oci_instance.instance_type
        name = f"{node_prefix}-{self.region}-{node_index}".lower()
        ssh_login_info = {
            "hostname": None,
            "user": oci_instance.user_name,
            "key_file": credentials.key_file,
            "extra_ssh_options": "-tt",
        }
>       super().__init__(
            name=name,
            parent_cluster=parent_cluster,
            ssh_login_info=ssh_login_info,
            base_logdir=base_logdir,
            node_prefix=node_prefix,
            dc_idx=dc_idx,
            rack=rack,
            after_
```
</details>

#### `unit.test_cluster_oci.test_private_dns_name_falls_back_to_reverse_dns_when_metadata_query_fails`
**Type:** FAILURE
**Message:** `TypeError: _base_node_init() got an unexpected keyword argument 'after_config'`
<details><summary>Traceback</summary>

```
mock_resolver = <MagicMock name='resolve_ip_to_dns' id='136917591463584'>

    @patch("sdcm.cluster_oci.resolve_ip_to_dns", return_value="node.internal")
    @patch("sdcm.cluster.BaseNode.__init__", new=_base_node_init)
    def test_private_dns_name_falls_back_to_reverse_dns_when_metadata_query_fails(mock_resolver) -> None:
        """Test that reverse DNS lookup is used when OCI metadata query raises an exception."""
>       oci_node = OciNode(_mock_cloud_instance(), MOCK_CREDENTIALS, MOCK_PARENT_CLUSTER)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

unit_tests/unit/test_cluster_oci.py:51: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <sdcm.cluster_oci.OciNode object at 0x7c869c990a50>
oci_instance = <Mock id='136917591463920'>
credentials = <Mock id='136917979654400'>
parent_cluster = <Mock id='136917979654736'>, node_prefix = 'node'
node_index = 1, base_logdir = None, dc_idx = 0, rack = 0, after_config = None

    def __init__(
        self,
        oci_instance: VmInstance,
        credentials,
        parent_cluster,
        node_prefix="node",
        node_index=1,
        base_logdir=None,
        dc_idx=0,
        rack=0,
        after_config=None,
    ):
        self.node_index = node_index
        self.dc_idx = dc_idx
        self.parent_cluster = parent_cluster
        self._instance = oci_instance
        self._instance_type = oci_instance.instance_type
        name = f"{node_prefix}-{self.region}-{node_index}".lower()
        ssh_login_info = {
            "hostname": None,
            "user": oci_instance.user_name,
            "key_file": credentials.key_file,
            "extra_ssh_options": "-tt",
        }
>       super().__init__(
            name=name,
            parent_cluster=parent_cluster,
            ssh_login_info=ssh_login_info,
            base_logdir=base_logdir,
            node_prefix=node_prefix,
            dc_idx=dc_idx,
            rack=rack,

```
</details>

#### `unit.test_cluster_oci.test_private_dns_name_prefers_reverse_dns_fqdn_when_metadata_returns_short_label`
**Type:** FAILURE
**Message:** `TypeError: _base_node_init() got an unexpected keyword argument 'after_config'`
<details><summary>Traceback</summary>

```
mock_resolver = <MagicMock name='resolve_ip_to_dns' id='136917591462240'>

    @patch("sdcm.cluster_oci.resolve_ip_to_dns", return_value="db-node-short.subnet.vcn.oraclevcn.com")
    @patch("sdcm.cluster.BaseNode.__init__", new=_base_node_init)
    def test_private_dns_name_prefers_reverse_dns_fqdn_when_metadata_returns_short_label(mock_resolver) -> None:
        """Test that reverse DNS FQDN is preferred over a short hostname label from metadata."""
>       oci_node = OciNode(_mock_cloud_instance(), MOCK_CREDENTIALS, MOCK_PARENT_CLUSTER)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

unit_tests/unit/test_cluster_oci.py:64: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <sdcm.cluster_oci.OciNode object at 0x7c869c991310>
oci_instance = <Mock id='136917591461904'>
credentials = <Mock id='136917979654400'>
parent_cluster = <Mock id='136917979654736'>, node_prefix = 'node'
node_index = 1, base_logdir = None, dc_idx = 0, rack = 0, after_config = None

    def __init__(
        self,
        oci_instance: VmInstance,
        credentials,
        parent_cluster,
        node_prefix="node",
        node_index=1,
        base_logdir=None,
        dc_idx=0,
        rack=0,
        after_config=None,
    ):
        self.node_index = node_index
        self.dc_idx = dc_idx
        self.parent_cluster = parent_cluster
        self._instance = oci_instance
        self._instance_type = oci_instance.instance_type
        name = f"{node_prefix}-{self.region}-{node_index}".lower()
        ssh_login_info = {
            "hostname": None,
            "user": oci_instance.user_name,
            "key_file": credentials.key_file,
            "extra_ssh_options": "-tt",
        }
>       super().__init__(
            name=name,
            parent_cluster=parent_cluster,
            ssh_login_info=ssh_login_info,
            base_logdir=base_logdir,
            node_prefix=node_prefix,
            dc_id
```
</details>

#### `unit.test_cluster_oci.test_private_dns_name_falls_back_to_private_ip_when_reverse_dns_missing`
**Type:** FAILURE
**Message:** `TypeError: _base_node_init() got an unexpected keyword argument 'after_config'`
<details><summary>Traceback</summary>

```
mock_resolver = <MagicMock name='resolve_ip_to_dns' id='136917591461568'>

    @patch(
        "sdcm.cluster_oci.resolve_ip_to_dns",
        side_effect=ValueError("Unable to resolve IP: [Errno 1] Unknown host"),
    )
    @patch("sdcm.cluster.BaseNode.__init__", new=_base_node_init)
    def test_private_dns_name_falls_back_to_private_ip_when_reverse_dns_missing(mock_resolver) -> None:
        """Test that private IP is returned when both metadata and reverse DNS fail."""
>       oci_node = OciNode(_mock_cloud_instance(), MOCK_CREDENTIALS, MOCK_PARENT_CLUSTER)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

unit_tests/unit/test_cluster_oci.py:79: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <sdcm.cluster_oci.OciNode object at 0x7c869cdbb5c0>
oci_instance = <Mock id='136917591460224'>
credentials = <Mock id='136917979654400'>
parent_cluster = <Mock id='136917979654736'>, node_prefix = 'node'
node_index = 1, base_logdir = None, dc_idx = 0, rack = 0, after_config = None

    def __init__(
        self,
        oci_instance: VmInstance,
        credentials,
        parent_cluster,
        node_prefix="node",
        node_index=1,
        base_logdir=None,
        dc_idx=0,
        rack=0,
        after_config=None,
    ):
        self.node_index = node_index
        self.dc_idx = dc_idx
        self.parent_cluster = parent_cluster
        self._instance = oci_instance
        self._instance_type = oci_instance.instance_type
        name = f"{node_prefix}-{self.region}-{node_index}".lower()
        ssh_login_info = {
            "hostname": None,
            "user": oci_instance.user_name,
            "key_file": credentials.key_file,
            "extra_ssh_options": "-tt",
        }
>       super().__init__(
            name=name,
            parent_cluster=parent_cluster,
            ssh_login_info=ssh_login_info,
            base_logdir=base_logdir,
            node_prefix=node_pre
```
</details>

#### `unit.test_cluster_oci.test_private_dns_name_falls_back_to_node_name_when_private_ip_missing`
**Type:** FAILURE
**Message:** `TypeError: _base_node_init() got an unexpected keyword argument 'after_config'`
<details><summary>Traceback</summary>

```
@patch("sdcm.cluster.BaseNode.__init__", new=_base_node_init)
    def test_private_dns_name_falls_back_to_node_name_when_private_ip_missing() -> None:
        """Test that node name is returned as last resort when private IP is also unavailable."""
>       oci_node = OciNode(_mock_cloud_instance(None), MOCK_CREDENTIALS, MOCK_PARENT_CLUSTER)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

unit_tests/unit/test_cluster_oci.py:91: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <sdcm.cluster_oci.OciNode object at 0x7c869cdbba80>
oci_instance = <Mock id='136917591461232'>
credentials = <Mock id='136917979654400'>
parent_cluster = <Mock id='136917979654736'>, node_prefix = 'node'
node_index = 1, base_logdir = None, dc_idx = 0, rack = 0, after_config = None

    def __init__(
        self,
        oci_instance: VmInstance,
        credentials,
        parent_cluster,
        node_prefix="node",
        node_index=1,
        base_logdir=None,
        dc_idx=0,
        rack=0,
        after_config=None,
    ):
        self.node_index = node_index
        self.dc_idx = dc_idx
        self.parent_cluster = parent_cluster
        self._instance = oci_instance
        self._instance_type = oci_instance.instance_type
        name = f"{node_prefix}-{self.region}-{node_index}".lower()
        ssh_login_info = {
            "hostname": None,
            "user": oci_instance.user_name,
            "key_file": credentials.key_file,
            "extra_ssh_options": "-tt",
        }
>       super().__init__(
            name=name,
            parent_cluster=parent_cluster,
            ssh_login_info=ssh_login_info,
            base_logdir=base_logdir,
            node_prefix=node_prefix,
            dc_idx=dc_idx,
            rack=rack,
            after_config=after_config,
        )
E       TypeError: _base_node_init() got an unexpected keyword argument 'after_config'

sdcm/cluster_oci.py:74: Type
```
</details>

#### `unit.test_cluster_oci.test_create_node_certificate_includes_short_and_fqdn_dns_names`
**Type:** FAILURE
**Message:** `TypeError: _base_node_init() got an unexpected keyword argument 'after_config'`
<details><summary>Traceback</summary>

```
mock_create_cert = <MagicMock name='create_certificate' id='136917591460560'>
mock_resolver = <MagicMock name='resolve_ip_to_dns' id='136917591460896'>

    @patch("sdcm.cluster_oci.resolve_ip_to_dns", return_value="db-node-short.subnet.vcn.oraclevcn.com")
    @patch("sdcm.cluster_oci.create_certificate")
    @patch("sdcm.cluster.BaseNode.__init__", new=_base_node_init)
    def test_create_node_certificate_includes_short_and_fqdn_dns_names(mock_create_cert, mock_resolver) -> None:
        """Test that node certificate includes both short hostname and FQDN as DNS names."""
>       oci_node = OciNode(_mock_cloud_instance("10.0.3.13"), MOCK_CREDENTIALS, MOCK_PARENT_CLUSTER)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

unit_tests/unit/test_cluster_oci.py:115: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <sdcm.cluster_oci.OciNode object at 0x7c869c6905f0>
oci_instance = <Mock id='136917591459888'>
credentials = <Mock id='136917979654400'>
parent_cluster = <Mock id='136917979654736'>, node_prefix = 'node'
node_index = 1, base_logdir = None, dc_idx = 0, rack = 0, after_config = None

    def __init__(
        self,
        oci_instance: VmInstance,
        credentials,
        parent_cluster,
        node_prefix="node",
        node_index=1,
        base_logdir=None,
        dc_idx=0,
        rack=0,
        after_config=None,
    ):
        self.node_index = node_index
        self.dc_idx = dc_idx
        self.parent_cluster = parent_cluster
        self._instance = oci_instance
        self._instance_type = oci_instance.instance_type
        name = f"{node_prefix}-{self.region}-{node_index}".lower()
        ssh_login_info = {
            "hostname": None,
            "user": oci_instance.user_name,
            "key_file": credentials.key_file,
            "extra_ssh_options": "-tt",
        }
>       super().__init__(
            name=name,
            parent_cluster=par
```
</details>

[Full build log](https://jenkins.scylladb.com/job/sct-github-PRs-scan/job/scylla-cluster-tests/job/PR-14453/2/)

--- a/sdcm/sct_runner.py
+++ b/sdcm/sct_runner.py
@@ -21,6 +21,7 @@ import logging
 import random
 import string
 import tempfile
+from concurrent.futures import ThreadPoolExecutor, as_completed
 from contextlib import suppress
 from enum import Enum
 from functools import cached_property
@@ -2067,6 +2068,72 @@ def _manage_runner_keep_tag_value(
         return sct_runner_info
 
 
+def _process_sct_runner(
+    sct_runner_info: SctRunnerInfo,
+    test_status: str,
+    test_runner_ip: str,
+    dry_run: bool,
+    force: bool,
+) -> bool:
+    """SSH-check, tag-manage, and optionally terminate a single SCT runner.
+
+    Returns True if the runner was terminated, False otherwise.
+    """
+    LOGGER.info("Managing SCT runner: %s in region: %s", sct_runner_info.instance_name, sct_runner_info.region_az)
+    cmd = 'cat /home/ubuntu/sct-results/latest/events_log/critical.log | grep "TestTimeoutEvent"'
+
+    if sct_runner_info.cloud_provider == "aws":
+        sct_runner_name = sct_runner_info.instance_name.split(" ")[0]
+    else:
+        sct_runner_name = sct_runner_info.instance_name
+
+    timeout_flag = False
+    # ssh_run_cmd currently only works on AWS, no point of wasting time
+    # trying to lookup on AWS instance from GCP/Azure
+    if sct_runner_info.public_ips and sct_runner_info.cloud_provider == "aws":
+        ssh_run_cmd_result = ssh_run_cmd(
+            command=cmd, test_id=sct_runner_info.test_id, node_name=sct_runner_name, force_use_public_ip=True
+        )
+        timeout_flag = bool(ssh_run_cmd_result.stdout) if ssh_run_cmd_result else False
+
+    utc_now = datetime.datetime.now(tz=datetime.timezone.utc)
+    LOGGER.info("UTC now: %s", utc_now)
+
+    if not dry_run and test_runner_ip:
+        _manage_runner_keep_tag_value(
+            test_status=test_status,
+            utc_now=utc_now,
+            timeout_flag=timeout_flag,
+            sct_runner_info=sct_runner_info,
+            dry_run=dry_run,
+        )
+
+    if not force and sct_runner_info.keep:
+        if sct_runner_info.keep_action != "terminate":
+            LOGGER.info("Skip %s because keep_action `keep_action' != `terminate'", sct_runner_info)
+            return False
+        if sct_runner_info.launch_time is None:
+            LOGGER.info("Skip %s because `launch_time' is not set", sct_runner_info)
+            return False
+        try:
+            if (utc_now - sct_runner_info.launch_time).total_seconds() < int(sct_runner_info.keep) * 3600:
+                LOGGER.info("Skip %s, too early to terminate", sct_runner_info)
+                return False
+        except ValueError as exc:
+            LOGGER.warning("Value of `keep' tag is invalid: %s", exc)
+
+    if dry_run:
+        LOGGER.info("Skip %s because of dry-run", sct_runner_info)
+        return False
+
+    try:
+        sct_runner_info.terminate()
+        return True
+    except Exception as exc:  # noqa: BLE001
+        LOGGER.warning("Exception raised during termination of %s: %s", sct_runner_info, exc)
+        return False
+
+
 def clean_sct_runners(
     test_status: str,
     test_runner_ip: str = None,
@@ -2077,9 +2144,6 @@ def clean_sct_runners(
     force: bool = False,
 ) -> None:
     sct_runners_list = list_sct_runners(backend=backend, test_runner_ip=test_runner_ip, user=user, test_id=test_id)
-    timeout_flag = False
-    runners_terminated = 0
-    end_message = ""
 
     if not dry_run and test_runner_ip and force:
         sct_runner_info = sct_runners_list[0]
@@ -2087,62 +2151,24 @@ def clean_sct_runners(
         LOGGER.info("Forcibly terminated runner: %s", sct_runner_info)
         return
 
-    for sct_runner_info in sct_runners_list:
-        LOGGER.info("Managing SCT runner: %s in region: %s", sct_runner_info.instance_name, sct_runner_info.region_az)
-        cmd = 'cat /home/ubuntu/sct-results/latest/events_log/critical.log | grep "TestTimeoutEvent"'
-
-        if sct_runner_info.cloud_provider == "aws":
-            sct_runner_name = sct_runner_info.instance_name.split(" ")[0]
-        else:
-            sct_runner_name = sct_runner_info.instance_name
-
-        # ssh_run_cmd currently only works on AWS, no point of wasting time
-        # trying to lookup on AWS instance from GCP/Azure
-        if sct_runner_info.public_ips and sct_runner_info.cloud_provider == "aws":
-            ssh_run_cmd_result = ssh_run_cmd(
-                command=cmd, test_id=sct_runner_info.test_id, node_name=sct_runner_name, force_use_public_ip=True
-            )
-
-            timeout_flag = bool(ssh_run_cmd_result.stdout) if ssh_run_cmd_result else False
-
-        utc_now = datetime.datetime.now(tz=datetime.timezone.utc)
-        LOGGER.info("UTC now: %s", utc_now)
-
-        if not dry_run and test_runner_ip:
-            _manage_runner_keep_tag_value(
-                test_status=test_status,
-                utc_now=utc_now,
-                timeout_flag=timeout_flag,
+    with ThreadPoolExecutor(max_workers=min(len(sct_runners_list), 32) or 1) as executor:
+        futures = [
+            executor.submit(
+                _process_sct_runner,
                 sct_runner_info=sct_runner_info,
+                test_status=test_status,
+                test_runner_ip=test_runner_ip,
                 dry_run=dry_run,
+                force=force,
             )
+            for sct_runner_info in sct_runners_list
+        ]
+        runners_terminated = sum(f.result() for f in as_completed(futures))
 
-        if not force and sct_runner_info.keep:
-            if sct_runner_info.keep_action != "terminate":
-                LOGGER.info("Skip %s because keep_action `keep_action' != `terminate'", sct_runner_info)
-                continue
-            if sct_runner_info.launch_time is None:
-                LOGGER.info("Skip %s because `launch_time' is not set", sct_runner_info)
-                continue
-            try:
-                if (utc_now - sct_runner_info.launch_time).total_seconds() < int(sct_runner_info.keep) * 3600:
-                    LOGGER.info("Skip %s, too early to terminate", sct_runner_info)
-                    continue
-            except ValueError as exc:
-                LOGGER.warning("Value of `keep' tag is invalid: %s", exc)
-
-        if dry_run:
-            LOGGER.info("Skip %s because of dry-run", sct_runner_info)
-            continue
-        try:
-            sct_runner_info.terminate()
-            runners_terminated += 1
-            end_message = f"Number of cleaned runners: {runners_terminated}"
-        except Exception as exc:  # noqa: BLE001
-            LOGGER.warning("Exception raised during termination of %s: %s", sct_runner_info, exc)
-            end_message = "No runners have been terminated"
-
-    LOGGER.info(end_message)
+    if runners_terminated:
+        LOGGER.info("Number of cleaned runners: %d", runners_terminated)
+    else:
+        LOGGER.info("No runners have been terminated")
 
 
 class AwsFipsSctRunner(AwsSctRunner):

